New wizards + modes

xmrdsc opened the pull request:
This PR is the result of rewriting and redesigning the wizards. Summary:

- All screens have been built up from scratch using [knueffelbunds designs](https://app.zeplin.io/project/5a0777492a92d8ac5beb3125/dashboard) as reference. 
- Language sidebar makes it possible to switch language at any time. Introduces a flag icon in the titlebar.
- Splash/Welcome shows upon first launch of the GUI.
- Access wallets from a 'most recently opened' list.
- Images are high DPI (Should look nice on retina).
- Wallet name/location inputs check for duplicate wallet files on-the-fly (as you can't overwrite existing wallets).
- Some input fields have regexp filters (restore height, subaddress lookahead).

**Edit:**
- Now also includes mode selection wizards and functionality.

Technical notes:

- Javascript logic related to opening/closing of wallets has been refactored and moved into `wizardController`.
- An effort has been made to use explicit variable names, i.e: `wizardController.variablename` instead of  merely `variable` (which is inherited via component hierarchy). This should in theory make the code more verbose and reveal where variables are actually coming from.
- `Layout.` used wherever possible. Still some 'mixing anchors/layouts' warnings though. Could not get rid of them.
- "recent wallets" functionality is implemented completely inside QML via `FolderListModel`, sorted on time modified.

Heads-ups for testers:

- Most files under `wizard/` have had their name changed which renders any existing translation mappings broken. I plan to write a Python script to (partially) migrate some translation strings to ease the pain for the translation workgroup.
- Test creating a wallet
  - Using different languages
  - Test restore height functionality
- Test if creating a wallet using Ledger works
  - Test restore height functionality
  - Test subaddress lookahead
- Test if you can break "recent wallets"
- Test performance of rotating globe on the welcome screen
- Test if dynamically updating translations work (except for the wizards)

## screenshots

![](https://i.imgur.com/e35VKs2.png)

![](https://i.imgur.com/QHrJDLI.png)

![](https://i.imgur.com/As8fMoR.png)

![](https://i.imgur.com/ntpui1A.png)

![](https://i.imgur.com/ZFz2c0t.png)

![](https://i.imgur.com/sgERn0E.png)

![](https://i.imgur.com/lTmFd9r.png)

![](https://i.imgur.com/p2zKEme.png)

![](https://i.imgur.com/JJoG4Zs.png)

![](https://i.imgur.com/BSftW0p.png)

![](https://i.imgur.com/OZabAwS.png)

selsta commented:
Things I’ve noticed so far:
- The wizard page is a little bit smaller than the normal GUI.
![width](https://user-images.githubusercontent.com/7697454/51384841-dac73280-1b1d-11e9-956a-e108486c4ce8.gif)

- Language sidebar isn’t usable if custom decorations are off, the icon still seems to get slightly displayed.
<img width="1042" alt="screenshot 2019-01-18 at 12 33 25" src="https://user-images.githubusercontent.com/7697454/51384660-4230b280-1b1d-11e9-9799-f20b3161c78a.png">

- Open a wallet from file -> Browse filesystem opens the GUI without a wallet open. If you press cancel inside the file dialog, the GUI stays open with no wallet file loaded.
![filedialog](https://user-images.githubusercontent.com/7697454/51385177-c20b4c80-1b1e-11e9-9662-0451e1c5f879.gif)

xmrdsc commented:
Thanks @selsta. I made some changes.


> The wizard page is a little bit smaller than the normal GUI.

Window size between the wizards and wallet view stays the same now.

> Language sidebar isn’t usable if custom decorations are off, the icon still seems to get slightly displayed.

Without custom decorations, the language can now be changed by going to `Settings->Layout`. In addition, there will be a button visible on the 'wizard home' screen to change the language.

> Open a wallet from file -> Browse filesystem opens the GUI without a wallet open. If you press cancel inside the file dialog, the GUI stays open with no wallet file loaded.

During opening of a wallet, the GUI will now stay in the wizard view.

> Language sidebar is empty for me (GUI and inside wizard).

Possibly related to your local setup, qmake should include these translations (via the `.pro` file).

In addition, I also rebased to latest master.

selsta commented on components/LanguageSidebar.qml:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    width: 240 * scaleRatio
+    height: parent.height - 50
```
```suggestion
    height: parent.height - (persistentSettings.customDecorations ? 50 : 0)
```

xmrdsc replied:
Thanks

selsta commented on components/LanguageSidebar.qml:
```diff
@@ -0,0 +1,159 @@
+
+        onStatusChanged: {
+            if(status === XmlListModel.Ready){
+                console.log("languages availible: ",count);
```
typo :P

xmrdsc replied:
Fixed :) Was from old code.

selsta commented on wizard/WizardHome.qml:
```diff
@@ -0,0 +1,221 @@
+                text: "Change language"
+
+                onClicked: {
+                    wizardController.wizardState = 'wizardLanguage';
```
```suggestion
                    wizardController.wizardState = 'wizardLanguage';
                    languageSidebar.open();
```

xmrdsc replied:
Nice one :)

selsta commented on components/LanguageSidebar.qml:
```diff
@@ -0,0 +1,159 @@
+                console.log("languages availible: ",count);
+                if(count === 1){
+                    console.log("Skipping language page until more languages are availible")
+                    wizard.switchPage(true);
```
```suggestion
                    if (wizardStateView.state = "wizardLanguage") {
                        wizard.switchPage(true);
                    }
```
I think this is more correct.

xmrdsc replied:
Thanks.

selsta commented:
I tried to debug the issue with languages not showing up without resizing but I’ve had no luck.

The only other improvement suggestion I have is some kind of scrollbar or something. Currently, if the window height isn’t high enough, there’s no way to move forward.

xmrdsc commented:
Note to self: integrate #1892

mmbyday commented on wizard/WizardCreateDevice1.qml:
```diff
@@ -0,0 +1,184 @@
+
+                    var written = wizardController.createWalletFromDevice();
+                    if(written){
+                        wizardStateView.state = "wizardHome";
```
```suggestion
                        wizardStateView.state = "wizardCreateWallet2";
```
Otherwise, wizard will not advance to next page.

xmrdsc replied:
Oh nice, you tested Ledger. I have yet to. Ill make these changes for the time being. Thanks.

mmbyday commented on wizard/WizardCreateDevice1.qml:
```diff
@@ -0,0 +1,184 @@
+import "../js/Wizard.js" as Wizard
+import "../components"
+import "../components" as MoneroComponents
+
```
```suggestion
import moneroComponents.Wallet 1.0

```

Fixes 
qrc:/wizard/WizardController.qml:342: ReferenceError: Wallet is not defined

mmbyday commented:
> > Language sidebar is empty for me (GUI and inside wizard).
> 
> Possibly related to your local setup, qmake should include these translations (via the `.pro` file).
> 

Also empty language sidebar for me, even with a clean build.

mmbyday commented:
Seems like there's a regression with https://github.com/monero-project/monero-gui/pull/1878

selsta commented on wizard/WizardNav.qml:
```diff
@@ -0,0 +1,116 @@
+        // Set `wizardProgress` width based on amount of progress dots
+        wizardProgress.width = 30 * menuNav.progressSteps;
+    }
+}
```
newline missing :P

xmrdsc replied:
:) fixed

mmbyday commented:
This seems broken.
* Test restore height functionality

Also, on the final summary page of the wizard, when creating a new wallet from seed restore, the restore height is no longer displayed. The old wizard would show restore height. Working as intended? Or just missing.

mmbyday commented on main.qml:
```diff
@@ -1891,6 +1954,14 @@ ApplicationWindow {
         passwordDialog.open();
     }
 
+    function changeWalletMode(mode){
```
Suggestion: Perhaps in the Layout/UI settings tab, allow users to change wallet modes.

mmbyday replied:
Nvm. I see the mode selection button on the welcome to monero screen. Nice :)

xmrdsc replied:
Still a good suggestion. I think we can think about this at a later point. Would be nice to be able to more easily switch modes (whilst you have a wallet opened).

mmbyday commented:
> I think my comment about a scrollbar is still valid, also the language bar still requires changing the window height to show up.
> 
> Will review more later.

Confirmed. While language bar is open but empty, resizing the window will cause the languages to appear.

mmbyday commented on wizard/WizardModeRemoteNodeWarning.qml:
```diff
@@ -0,0 +1,206 @@
+                Layout.fillWidth: true
+
+                Text {
+                    text: qsTr("This mode is ideal for managing small amounts of money. You have access to basic features for making and managing transactions. It will automatically connect to the Monero network so you can start using Monero immediately.") + translationManager.emptyString
```
```suggestion
                    text: qsTr("This mode is ideal for managing small amounts of monero. You have access to basic features for making and managing transactions. It will automatically connect to the Monero network so you can start using Monero immediately.") + translationManager.emptyString
```

better IMO :)

xmrdsc replied:
Oh, good catch.

mmbyday commented on wizard/WizardModeSelection.qml:
```diff
@@ -0,0 +1,134 @@
+                btnNext.visible: false
+                progressSteps: 0
+
+                onPrevClicked: {
```
Suggestion: This Mode Selection screen could use a "previous" button, along with the change language button.

xmrdsc replied:
Where would previous bring you?

xmrdsc replied:
I guess back to the main menu?

xmrdsc replied:
It would add some complexity to determine if an user already has already chosen a mode, in which case you can show the previous button but not on the first run. I would prefer to leave it like it is.

rbrunner7 replied:
One more vote for a *Back to menu* button on the *Mode selection* screen: If you arrive here from the menu it's impossible to back out without making a choice, with the added difficulty that right now this screen does not indicate which mode is the currently active one.

mmbyday commented on wizard/WizardHome.qml:
```diff
@@ -0,0 +1,246 @@
+
+                MoneroComponents.StandardButton {
+                    small: true
+                    text: qsTr("Mode selection") + translationManager.emptyString
```
```suggestion
                    text: qsTr("Change wallet mode") + translationManager.emptyString
```
For better button text consistency

xmrdsc replied:
:+1:

mmbyday commented on wizard/WizardHome.qml:
```diff
@@ -0,0 +1,246 @@
+                }
+            }
+
+            RowLayout {
```
Perhaps these two buttons would look better centered.

xmrdsc replied:
Don't think it will look OK since it is a change from the 'far-left' & 'far-right' button positions we have currently. In addition, there is no guarantee that there are 2 buttons (depending on custom decorations).

xmrdsc replied:
Feel free to PR for it after this is merged though if you think it's important.

mmbyday commented on wizard/WizardLanguage.qml:
```diff
@@ -0,0 +1,253 @@
+                Layout.minimumWidth: 150 * scaleRatio
+                text: "Continue"
+
+                onClicked: {
```
Is this extra click on the continue button required. If I click a language, that should be good to move on.

xmrdsc replied:
I think this is pretty confusing in terms of UX. Clicking language in the sidebar should not do anything more than (dynamically) switching the language it is currently displaying.

In addition, I think the complexity of code needed does not justify the goal.

mmbyday commented on components/LanguageSidebar.qml:
```diff
@@ -0,0 +1,155 @@
+                    opacity: 0.75
+                    visible: true
+
+                    MouseArea {
```
A mouseover effect, like that on the accounts page's listview, would be nice. 
i.e.                                 
```
                                onEntered: {
                                    delegateTable.color = "#26FFFFFF"
                                }
                                onExited: {
                                    delegateTable.color = "transparent"
                                }
```

xmrdsc replied:
The background of those items is actually a picture. However, I was able to simulate a hover effect playing around with `opacity`. Looks better indeed.

mmbyday commented on wizard/WizardLanguage.qml:
```diff
@@ -0,0 +1,253 @@
+            font.pixelSize: 12 * scaleRatio
+            font.family: MoneroComponents.Style.fontRegular.name
+            color: MoneroComponents.Style.defaultFontColor
+            text: Version.GUI_VERSION + " (Qt " + qtRuntimeVersion + ")"
```
Perhaps some pre-text. Otherwise this text looks odd without context.
```suggestion
            text: "Wallet version: " + Version.GUI_VERSION + " (Qt " + qtRuntimeVersion + ")"
```

xmrdsc replied:
That's not odd. That's just how software usually display their versions in splashes and/or intros.

mmbyday commented on wizard/WizardCreateWallet4.qml:
```diff
@@ -0,0 +1,86 @@
+
+                    wizardStateView.state = "wizardCreateWallet3";
+                }
+                onNextClicked: {
```
Scenario: restore wallet, open wallet, logout of wallet, restore wallet, then previous mnemonic still visible. Perhaps clear sensitive fields at this point.

xmrdsc replied:
:+1:

xmrdsc commented on wizard/WizardController.qml:
```diff
@@ -0,0 +1,479 @@
+
+            delegate: StackViewDelegate {
+                pushTransition: StackViewTransition {
+                    // PropertyAnimation {
```
Yes, I don't see why not.

selsta replied:
Why is it commented out in the first place?

xmrdsc replied:
I've removed the comment. On second thought the animations look pretty sweet.

xmrdsc commented on wizard/WizardAskPassword.qml:
```diff
@@ -0,0 +1,264 @@
+          strength = 100;
+
+        // privacyLevel component uses 1..100 scale
+        strength = Wizard.mapScope(1, 100, 1, 100, strength);
```
?

selsta replied:
This function does nothing if the input and output scale is both 1, 100.

selsta replied:
12:13 <selsta> dsc_: If the password strength is fine, remove the mapscope function IMO.

xmrdsc commented on wizard/WizardController.qml:
```diff
@@ -0,0 +1,479 @@
+        if (success) {
+            wizardController.m_wallet = wallet;
+            wizardController.walletOptionsIsRecoveringFromDevice = true;
+            wizardController.tmpWalletFilename = tmp_wallet_filename;
```
Fixed

xmrdsc commented on wizard/WizardRestoreWallet1.qml:
```diff
@@ -0,0 +1,295 @@
+                            anchors.margins: 8 * scaleRatio
+                            anchors.leftMargin: 10 * scaleRatio
+                            font.family: MoneroComponents.Style.fontRegular.name
+                            text: qsTr("Enter your 25 (or 24) word mnemonic seed") + translationManager.emptyString
```
Currently: https://github.com/monero-project/monero-gui/blob/dc7a95c8d18d39268f1bc8d87701bfba90a91bbf/wizard/WizardMemoTextInput.qml#L45

Probably checksum.

selsta replied:
Ok, I think we should remove mentions of 24 word seeds but not in this PR.

xmrdsc commented on wizard/WizardWalletInput.qml:
```diff
@@ -0,0 +1,112 @@
+            if(walletLocation === "") return false;
+
+            var exists = Wizard.walletPathExists(walletLocation.text, walletName.text, isIOS, walletManager);
+            if(!exists && walletLocation.error === false) return true;
```
Please elaborate

selsta replied:
if (x) return true;
else return false;

is the same as

return x;

selsta replied:
ping on this? It’s not a bug but it simplifies the code. Writing the return value explicit as you did isn’t usual.

xmrdsc replied:
Fixed

xmrdsc commented:
Thanks for review... Some fixes/changes:

- Fixed empty language sidebar with Qt >5.7
- Hover effect language sidebar similiar to accounts/receive table
- Fixed creating wallet with Ledger
- Revamp WizardSummary.qml, made more modular. Actually readable code now.
- Clear fields on reset for 'restore from device'

selsta commented on wizard/WizardController.qml:
```diff
@@ -0,0 +1,482 @@
+
+        var wallet = walletManager.createWalletFromDevice(tmp_wallet_filename, "", nettype, deviceName, restoreHeight, subaddressLookahead);
+
+        var success = wallet.status === Wallet.Status_Ok;
```
Requires `import moneroComponents.Wallet 1.0`.

xmrdsc replied:
:+1:

selsta commented on wizard/WizardController.qml:
```diff
@@ -0,0 +1,482 @@
+            wizardController.walletOptionsRestoreHeight = wizardController.m_wallet.walletCreationHeight;
+        } else {
+            console.log(wallet.errorString)
+            walletErrorDialog.text = wallet.errorString;
```
`walletErrorDialog` is undefined as far as I can see.

xmrdsc replied:
:+1:

selsta commented:
Bugs I’ve found:

- Simple mode, remote node: Creating a new wallet skips the node page, but pressing `Previous` on the `You're all set up` page sends you to the node page.
- Advanced mode: It will show a random remote node address on the `You're all set up` page even if you select local node only.
- If you create a wallet from device and press the `Previous` button, you will get to wrong page.
- Creating wallet from device is missing the dots.

selsta commented on wizard/WizardSummary.qml:
```diff
@@ -0,0 +1,98 @@
+        value: walletOptionsName
+    }
+
+    WizardSummaryItem {
```
Debug leftover?

xmrdsc replied:
Thanks :)

selsta commented on js/Utils.js:
```diff
@@ -87,3 +87,22 @@ function ago(epoch) {
+
+function netTypeToString(){
+    var nettype = appWindow.persistentSettings.nettype;
+    return nettype == NetworkType.TESTNET ? qsTr("Testnet") : nettype == NetworkType.STAGENET ? qsTr("Stagenet") : qsTr("Mainnet");
```
`qrc:/js/Utils.js:94: ReferenceError: NetworkType is not defined`

xmrdsc replied:
Fixed

xmrdsc commented on wizard/WizardSummary.qml:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    WizardSummaryItem {
+        visible: persistentSettings.remoteNodeAddress !== ""
```
I did `visible: persistentSettings.remoteNodeAddress !== "" && appWindow.walletMode == 0` instead, same for the bootstrap item.

selsta replied:
Yep, this seems fixed now.

xmrdsc commented on wizard/WizardSummary.qml:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    WizardSummaryItem {
+        visible: persistentSettings.bootstrapNodeAddress !== ""
```
No mistake here, works as intended.

selsta replied:
Like I said on IRC, this doesn’t look right. Nettype has nothing to do with bootstrapNodeAddress. I’d say it should be always visible or only visible if the nettype is testnet/stagenet.

xmrdsc replied:
Sorry, you're right. Fixed.

xiphon commented:
The PR is too big, I can't review it.
Have couple remarks:
1. For obvious reason, i don't like the idea to fetch remote nodes lsit from  Who controls "https://autonode.xmr.pm/"

xiphon commented:
The PR is too big, I can't review it.
Have couple general remarks:
1. For obvious reason, i don't like the idea to hardcode `https://autonode.xmr.pm/` URL to fetch remote nodes list from.
Given that most Monero GUI users will use default hardcoded url and Simple mode, this change makes them to blindly trust to some person controlling `https://autonode.xmr.pm/`.
Such a kind of centralization introduces single point of failure, heavily affects most Monero GUI users' privacy.
2. I don't get the idea behind the rotating globe welcoming screen. Why do we need it? What are the pros in terms of UX?

selsta commented:
We are getting close to merging this :)

I think this should still get fixed though:

> If you create a wallet from device and press the `Previous` button, you will get to wrong page.

https://user-images.githubusercontent.com/7697454/53013705-b75a1500-3446-11e9-835b-d580665d7543.gif

xmrdsc commented:
Thanks selsta;

- Removed mapScore (password view)
- Fixed 'Previous' button while restoring from device
- Fixed network type visibility on the 'All finished page'

@xiphon regarding rotating globe; perhaps talk with kneuffelbund on IRC and see what he thinks. Maybe he can come up with something. Removing it all together is also fine by me but would rather do that in a separate PR where we can discuss the starting screen more in depth.

selsta commented on wizard/WizardDaemonSettings.qml:
```diff
@@ -26,171 +26,142 @@
                 id: bootstrapNodeEdit
+                Layout.minimumWidth: 300 * scaleRatio
+
+                //labelText: qsTr("Bootstrap node (leave blank if not wanted)") + translationManager.emptyString
```
I think this isn’t supposed to stay in?

selsta commented on wizard/WizardNavProgressDot.qml:
```diff
@@ -1,21 +1,21 @@
-// Copyright (c) 2014-2018, The Monero Project
-//
+// Copyright (c) 2014-2019, The Monero Project
+// 
```
Stray spaces in the following lines look unintended.
